src/herder/HerderImpl.cpp

```diff
@@ -898,18 +934,8 @@ HerderImpl::getCurrentLedgerSeq() const
 {
     uint32_t res = mLedgerManager.getLastClosedLedgerNum();
 
-    if (mHerderSCPDriver.trackingSCP() &&
-        res < mHerderSCPDriver.trackingSCP()->mConsensusIndex)
-    {
-        res = static_cast<uint32_t>(
-            mHerderSCPDriver.trackingSCP()->mConsensusIndex);
-    }
-    if (mHerderSCPDriver.lastTrackingSCP() &&
-        res < mHerderSCPDriver.lastTrackingSCP()->mConsensusIndex)
-    {
-        res = static_cast<uint32_t>(
-            mHerderSCPDriver.lastTrackingSCP()->mConsensusIndex);
-    }
+    res = std::max(res, trackingConsensusLedgerIndex());
```

Comment:
this is not equivalent: old code was checking that tracking was set, new code would crash (as it tries to access index while it's not tracking)

Reply:
(I expect a refactor commit to not make functional changes)

Reply:
> this is not equivalent: old code was checking that tracking was set, new code would crash (as it tries to access index while it's not tracking)

I did not mean this to be equivalent. In this PR, I am making the "tracking" invariant stronger, i.e. it is always set on startup (see the change to `restoreSCPState`). With this guarantee, it becomes invalid to call these functions while in booting state (as we saw in some tests that wouldn't start the app properly, for example). Like I said before, I had to squash commits because the rebase turned out to be pretty painful (there used to be individual commits with functional/non-functional changes), but I guess I can attempt it again to make it easier to review.